java-checks/src/main/java/org/sonar/java/checks/NoTestInTestClassCheck.java

```diff
@@ -68,60 +69,74 @@ public boolean apply(SymbolMetadata.AnnotationInstance input) {
 
   @Override
   public void visitNode(Tree tree) {
-    CompilationUnitTree cut = (CompilationUnitTree) tree;
-    for (Tree typeTree : cut.types()) {
-      if (typeTree.is(Kind.CLASS)) {
-        ClassTree classTree = (ClassTree) typeTree;
-        if (!ModifiersUtils.hasModifier(classTree.modifiers(), Modifier.ABSTRACT)) {
-          if (classTree.symbol().metadata().isAnnotatedWith("org.testng.annotations.Test")) {
-            checkTestNGmembers(classTree);
-          } else {
-            checkJunit3TestClass(classTree);
-            checkJunit4TestClass(classTree);
-          }
+    if (hasSemantic()) {
+      CompilationUnitTree cut = (CompilationUnitTree) tree;
+      for (Tree typeTree : cut.types()) {
+        if (typeTree.is(Kind.CLASS)) {
+          checkClass((ClassTree) typeTree);
         }
       }
     }
   }
 
-  private void checkTestNGmembers(ClassTree classTree) {
-    for (Tree member : classTree.members()) {
-      if (member.is(Kind.METHOD)) {
-        Symbol.MethodSymbol symbol = ((MethodTree) member).symbol();
-        if (symbol.isPublic() && !symbol.isStatic()) {
-          return;
-        }
+  private void checkClass(ClassTree classTree) {
+    if (!ModifiersUtils.hasModifier(classTree.modifiers(), Modifier.ABSTRACT)) {
+      JavaSymbol.TypeJavaSymbol symbol = (JavaSymbol.TypeJavaSymbol) classTree.symbol();
+      Iterable<Symbol> members = getAllMembers(symbol);
+      IdentifierTree simpleName = classTree.simpleName();
+      if (classTree.symbol().metadata().isAnnotatedWith("org.testng.annotations.Test")) {
+        checkTestNGmembers(simpleName, members);
+      } else {
+        checkJunit3TestClass(simpleName, symbol, members);
+        checkJunit4TestClass(simpleName, symbol, members);
       }
     }
-    reportIssue(classTree.simpleName(), "Add some tests to this class.");
   }
 
-  private void checkJunit3TestClass(ClassTree tree) {
-    if (tree.symbol().type().isSubtypeOf("junit.framework.TestCase")) {
-      checkMethods(tree, false);
+  private void checkTestNGmembers(IdentifierTree identifierTree, Iterable<Symbol> members) {
+    for (Symbol member : members) {
+      if (member.isMethodSymbol() && member.isPublic() && !member.isStatic()) {
+        return;
+      }
     }
+    reportIssue(identifierTree, "Add some tests to this class.");
   }
 
-  private void checkJunit4TestClass(ClassTree tree) {
-    IdentifierTree name = tree.simpleName();
-    if (name != null && name.name().endsWith("Test")) {
-      checkMethods(tree, true);
+  private void checkJunit3TestClass(IdentifierTree simpleName, JavaSymbol.TypeJavaSymbol symbol, Iterable<Symbol> members) {
```

Comment:
@mpaladin I would prefer using a more meaningful name for the identifier tree from the class name.